src/util.c

```diff
@@ -118,6 +120,39 @@ gboolean CacheUtilHashalgFromString(const gchar *value, GChecksumType *result) {
 
 
 /**
+ * CacheUtilReadShadowFile:
+ * @path: Shadow file path.
+ * @username: Username to read hashed password for.
+ * @error Error return location or #NULL. 
+ *
+ * Returns: GByte hash for user, or #NULL on error. 
+ */
+
+GBytes *CacheUtilReadShadowFile(const gchar *path, const gchar *username, GError **error) {
+  struct spwd *stmpent = NULL;
+  
+
+  FILE *shadowfile = fopen(path, "r");
+  if (shadowfile == NULL) {
+    g_set_error(error, UTIL_ERROR, UTIL_ERROR_NO_OPEN_FILE, "Could not open shadow file.");
```

Comment:
Use errno to say more about what actually happened. See the foo_open_file example in https://developer.gnome.org/glib/stable/glib-Error-Reporting.html